tar/bsdtar.c

```diff
@@ -122,10 +125,72 @@ static void		 version(void);
 	(ARCHIVE_EXTRACT_SECURE_SYMLINKS		\
 	 | ARCHIVE_EXTRACT_SECURE_NODOTDOT)
 
+static struct bsdtar bsdtar_storage;
+
+static struct bsdtar *
+bsdtar_init(void)
+{
+	struct bsdtar * bsdtar = &bsdtar_storage;
+
+	memset(bsdtar, 0, sizeof(*bsdtar));
+
+	/* Set pointers to NULL for non-0 NULL platforms. */
+	bsdtar->tapenames = NULL;
+	bsdtar->homedir = NULL;
+	bsdtar->cachedir = NULL;
+	bsdtar->pending_chdir = NULL;
+	bsdtar->names_from_file = NULL;
+	bsdtar->modestr = NULL;
+	bsdtar->configfiles = NULL;
+	bsdtar->archive = NULL;
+	bsdtar->progname = NULL;
+	bsdtar->argv = NULL;
+	bsdtar->optarg = NULL;
+	bsdtar->write_cookie = NULL;
+	bsdtar->chunk_cache = NULL;
+	bsdtar->argv_orig = NULL;
+	bsdtar->delopt = NULL;
+	bsdtar->delopt_tail = NULL;
+	bsdtar->diskreader = NULL;
+	bsdtar->resolver = NULL;
+	bsdtar->gname_cache = NULL;
+	bsdtar->buff = NULL;
+	bsdtar->matching = NULL;
+	bsdtar->security = NULL;
+	bsdtar->uname_cache = NULL;
+	bsdtar->siginfo = NULL;
+	bsdtar->substitution = NULL;
+
+	/* We don't have bsdtar->progname yet, so we can't use bsdtar_errc. */
+	if (atexit(bsdtar_atexit)) {
+		fprintf(stderr, "tarsnap: Could not register atexit.\n");
```

Comment:
FWIW, I think it's cleaner to initialize `bsdtar->progname` inside `bsdtar_init` -- IMO, the returned value from an "init" function should be ready to use.  But I agree with sticking with this version for now, since I'm sick of seeing lots of junk in valgrind when I test other things.  :)